### PR TITLE
Add sum gradient, fix div gradient, add first autoencoder.

### PR DESCRIPTION
**Pull request by whilo**
I am not sure whether I am doing the gradients right, but the div gradient is not right I think (it also has no test). I am not sure what to do about the transpose yet, as traditional autoencoders use transposed weight matrices. The VAEs don't do so, but I still think it would be good to implement a normal autoencoder first. The error is still also high, but it is the minimal architecture (hidden layer size 2), to encode 4 different inputs, so I am not sure whether I am doing something wrong.

### CODE REVIEW

**Comment by kieranbrowne** [comment]
Hi thanks again for the PR. The basic autoencoder looks good. It broke because of a change I just made to gradient-descent, but I'll fix that up now.
Regarding Transpose. That's a tough one. There's a lot of tf ops that aren't obviously differentiable.
There's obviously a solution because it works in the python version, but their code isn't so easy to read.
Were you saying auto differentiation is your area of research, aside from VAEs?

**Comment by kieranbrowne** [comment]
It would be great to move your basic autoencoder into an example as well. I think people will find it useful.

**Comment by whilo** [comment]
My master thesis is about a combination of Bayesian Program Learning (or classical graphical models, we will see) with neural networks. I just need a research friendly differentiable pipeline in Clojure as I want to combine it with Anglican. I don't want to reimplement too much right now though. In the long run I can contribute, but I need something which I can use for my research. So I am really grateful that you started to wrap tensorflow! 

Regarding transposition. In backpropagation the update of the matrix is just the outer product of 
`\delta_{back} \x_{forward}^T` (no math support in github yet). So you just have to transpose this outer product for the gradient with respect to the matrix and project the error back through this transposed matrix. I do not understand how the registered-gradients work fully yet. Maybe you can open a gitter room (or something similar), so we can discuss this?

Btw. most people in our group prefer (Py)Torch now (they used TensorFlow before), because its graph is dynamic and values can be intercepted in normal Python control flow. I think it would be nice to have something like this, but implemented in Clojure. Maybe Torch is more portable since it is written in Lua. Having said this, TensorFlow is fine for me right now and a good base to kick off a pipeline for Clojure. I am just trying to understand what the tradeoffs are and what I would like to have.

**Comment by kieranbrowne** [comment]
Sounds cool. I'm doing a PhD with neural networks at the moment. Good fun.

Like you it seems, I'm not really in a position to be reimplementing autodiff although down the line, that seems like a must for clojure.
I've been mucking around with gradients again over the past week, and I'm pretty convinced that defining a gradient for every tf operation is not feasible.
I also came across [this](https://www.tensorflow.org/extend/language_bindings#gradients_functions_and_control_flow). Basically there will be no gradient computation for languages other than python in the near future.

So I've tried a new approach, from the whole "code is data" thing, and I'm parsing the core tensorflow C files to extract the gradients. It's already been pretty successful; i've managed to get a number of gradients including tanh, sqrt, sigmoid, sine, div and neg for free. But needs work in two key areas. Firstly my regexes for finding the key parts of the C code aren't working for all ops at the moment and second, there are some funky shorthands particularly in operation attributes which I need to parse. But I think its the go for now. The work that gets done should have better than linear results. And plus it'll solve gradients that I have literally no idea how to solve like conv2d.

My feeling about the Tensorflow vs. other frameworks debate is that TensorFlow might not be the best in every way, but its still the safest choice for the future. It has the biggest user share and support from Google, so it's only going to gain strength. Google also seems really keen to make it truly cross language and cross platform (even Android).

**Comment by whilo** [comment]
Cool, what are you doing your PhD on? I think the C parsing is a pragmatic approach which I would support for now. Having numerically good gradients will help along the way no matter what framework we use. What is the licencing story then?

I think while TensorFlow will have a lot of support, it won't be the only framework and backpropagation pipelines are fairly similar. So I think once you start building the backprop graph in Clojure anyway, one could just make the numeric operations polymorphic and plug different backends (including pure Clojure ones).
They mention on the page you linked that they will provide cross-platform support when the C API is ready. But stepping back from just getting something to work, I think one should keep in mind that Lisp was invented to do symbolic computations including symbolic differentiation and figure out the big picture a bit. There are already implementations like in SICMutils, so instead of just getting something to work, I would like to know how it should be done the Lisp way. If we start to implement symbolic differentiation for backpropagation, I think it would be reasonable to at least not just wrap tensorflow in the longer run, because it is the most important feature of tensorflow. The rest is mostly about wrapping LAPACK, doing some efficient convolution kernels for the GPU and non-linearities. I think the forward pass (what the tensorflow API is providing right now) is by far not as difficult if you have good LAPACK wrappers (like neanderthal). The convolutions are not easy to do efficiently on the GPU though, but one can probably port some stuff from cortex. Tensorflow is very stiff because it is compiled AOT and you cannot interrupt its control flow from the host language easily. That is why PyTorch is getting a lot of traction in my impression recently.
Do you have some requirements a symbolic differentiation framework should have for you? Maybe we can form a roadmap.

**Comment by whilo** [comment]
I just found a very good overview which clarified for me the distinction between symbolic and automatic differentiation, although I have still to understand most of the subtleties: http://alexey.radul.name/ideas/2013/introduction-to-automatic-differentiation/ They mention http://autodiff.org/ and quite a bit of literature. It seems scmutils is already doing the right thing, so maybe it is really usable. It is GPL atm. though, which can be a serious problem.

**Comment by kieranbrowne** [comment]
Hey,
My PhD is on accidental culture that can be learned by neural networks and other ML. E.g.  http://science.sciencemag.org/content/356/6334/183 
Not sure about the licensing story, but TF/Google almost certainly maintains copyright.
I don't really have explicit needs for a differentiation framework, the tensorflow wrapping stuff is kindof a distraction from my thesis when I could just use python, but it's helped me to understand how these things work much more deeply. 
Thanks for the links, I'll give them a read.
I think autodiff is definitely the way to go in the long run. Having said that, I don't know how much time you have remaining on your thesis, but we could probably spend the whole time mucking around with this stuff, so it might be better to use a python lib in the mean time.

**Comment by whilo** [comment]
That is cool. Yes, the bias in the data can be a big problem. Right, AD is fairly interesting, I found a good survey paper for auto-differentiation as well https://arxiv.org/abs/1502.05767. I agree, I will probably use Python, but I think your tensorflow approach could also work for not too fancy neural networks with a bit of effort. I really didn't want to demotivate you. I think Clojure would need a concerted effort towards scientific computing in general though. It has potential, but it is fairly weak and I don't see people really acknowledging it. There are talks about it since 2010, but most things never really took off (Incanter, core.matrix, neanderthal, Anglican, ...). There are different conference talks were people praise Clojure for machine learning, but honestly, I don't think it is really helpful unless you implement most stuff on your own and you have fairly small problems. Do you think such an effort would be worthwhile? 

We would need to define a concrete plan towards a good scientific computing environment and try to find parts of the Clojure community to commit to it.
